### PR TITLE
Cleanup the compiler instrumentation

### PR DESCRIPTION
**Pull request by mehrdad2m**
**Context:**
Compiler instrumentation is printing info after each call to a pass which makes the output messy and hard to read.

**Description of the Change:**
changed the instrumentation such that it only prints stats after a pipeline in favour of cleaner output. It is still prossible to get the more detailed output by enabling the verbose option in qjit function.

**Benefits:**
Cleaner output for instrumentation.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**

### CODE REVIEW

File: mlir/lib/Driver/CompilerDriver.cpp [hunk 1]
```diff
@@ -742,32 +756,32 @@ LogicalResult QuantumDriverMain(const CompilerOptions &options, CompilerOutput &
...
+            o2PassesTiming.stop();
 
             TimingScope enzymePassesTiming = llcTiming.nest("Enzyme passes");
             if (failed(timer::timer(runEnzymePasses, "runEnzymePasses", /* add_endl */ false,
                                     options, llvmModule, output))) {
                 return failure();
             }
-            enzymePassesTiming.stop();
             catalyst::utils::LinesCount::Module(*llvmModule.get());
+            enzymePassesTiming.stop();
```
Comment (dime10) [inline mlir/lib/Driver/CompilerDriver.cpp:784]: I don't think counting lines should be included in a runtime measurement 🤔

**Review by dime10 — COMMENTED** [review]
Thanks for the update, I'll try it locally before finishing the review :)

File: mlir/lib/Driver/CompilerDriver.cpp [hunk 2]
```diff
@@ -742,32 +756,32 @@ LogicalResult QuantumDriverMain(const CompilerOptions &options, CompilerOutput &
...
-        catalyst::utils::LinesCount::Module(*llvmModule.get());
-
-        TimingScope coroLLVMPassesTiming = llcTiming.nest("LLVM coroutine passes");
-        if (options.asyncQnodes &&
-            failed(timer::timer(runCoroLLVMPasses, "runCoroLLVMPasses", /* add_endl */ false,
-                                options, llvmModule, output))) {
-            return failure();
+        if (options.asyncQnodes) {
+            TimingScope coroLLVMPassesTiming = llcTiming.nest("LLVM coroutine passes");
+            if (failed(timer::timer(runCoroLLVMPasses, "runCoroLLVMPasses", /* add_endl */ false,
```
Comment (dime10) [inline mlir/lib/Driver/CompilerDriver.cpp:761]: Are these two different timing systems?

**Review by dime10 — APPROVED** [review]
Nice, seems to work well!

File: mlir/lib/Driver/CompilerDriver.cpp (hunk 2 above)
Comment (mehrdad2m) [reply mlir/lib/Driver/CompilerDriver.cpp:761]: Yes the timer::timer is for the frontend instrumentation, and the TimingScope one is for mlir-opt internal timing which enables us to use --mlir-timing option with catalyst-cli which would print a similar time breakdown.

**Comment by josh146** [comment]
@mehrdad2m if the instrumentation functionality/output has changed, do you mind updating

- https://docs.pennylane.ai/projects/catalyst/en/latest/code/api/catalyst.debug.instrumentation.html
- https://docs.pennylane.ai/projects/catalyst/en/latest/dev/debugging.html#profiling-and-instrumentation

File: mlir/lib/Driver/CompilerDriver.cpp (hunk 2 above)
Comment (dime10) [reply mlir/lib/Driver/CompilerDriver.cpp:761]: Oh neat, I'll have to try it at some point!

**Comment by dime10** [comment]
> @mehrdad2m if the instrumentation functionality/output has changed, do you mind updating
> 
> * https://docs.pennylane.ai/projects/catalyst/en/latest/code/api/catalyst.debug.instrumentation.html
> * https://docs.pennylane.ai/projects/catalyst/en/latest/dev/debugging.html#profiling-and-instrumentation

I think it's just restoring behaviour from before the compiler-cli change.

**Comment by mehrdad2m** [comment]
> > @mehrdad2m if the instrumentation functionality/output has changed, do you mind updating
> > 
> > * https://docs.pennylane.ai/projects/catalyst/en/latest/code/api/catalyst.debug.instrumentation.html
> > * https://docs.pennylane.ai/projects/catalyst/en/latest/dev/debugging.html#profiling-and-instrumentation
> 
> I think it's just restoring behaviour from before the compiler-cli change. But we could maybe add mention of the verbose option?

Yes, but the changes from this PR are only relevant if `detailed` option is true. There is no example in the documentation with `detailed` being true. unless we want to add an example with `detailed` being true?

File: mlir/lib/Driver/CompilerDriver.cpp (hunk 1 above)
Comment (mehrdad2m) [reply mlir/lib/Driver/CompilerDriver.cpp]: Thanks for catching. Idk why I did that 🫠 . Fixed it

File: frontend/test/lit/test_instrumentation_console.py [hunk 3]
```diff
@@ -79,6 +75,11 @@ def circuit(weights):
 # CHECK-NEXT: [DIAGNOSTICS] Running device_release
 # CHECK-SAME:   walltime: {{[0-9\.]+}} ms{{\s*}} cputime: {{[0-9\.]+}} ms
 # CHECK:      [DIAGNOSTICS] > Total run
+# COM: As the below output is generated by catalyst-cli checking the corect order may cause falky results.
+# CHECK: [DIAGNOSTICS] Running parseMLIRSource
+# CHECK-SAME:   walltime: {{[0-9\.]+}} ms{{\s*}} cputime: {{[0-9\.]+}} ms{{\s*}} programsize: {{[0-9]+}} lines
+# CHECK: [DIAGNOSTICS] Running {{[a-zA-Z]+}}Pass
+# CHECK-SAME:   walltime: {{[0-9\.]+}} ms{{\s*}} cputime: {{[0-9\.]+}} ms{{\s*}} programsize: {{[0-9]+}} lines
```
Comment (dime10) [inline frontend/test/lit/test_instrumentation_console.py:82]: Does that mean the output will now show out of order for the user or is this a rare occurrence?

Comment (mehrdad2m) [reply frontend/test/lit/test_instrumentation_console.py:82]: no It shows fine for user, but for some reason when piping the output through FileCHECK it messes up the order.

File: frontend/test/lit/test_instrumentation_console.py [hunk 4]
```diff
@@ -79,6 +75,11 @@ def circuit(weights):
 # CHECK-NEXT: [DIAGNOSTICS] Running device_release
 # CHECK-SAME:   walltime: {{[0-9\.]+}} ms{{\s*}} cputime: {{[0-9\.]+}} ms
 # CHECK:      [DIAGNOSTICS] > Total run
+# COM: As the below output is generated by catalyst-cli checking the corect order may cause falky results.
```
Comment (joeycarter) [inline frontend/test/lit/test_instrumentation_console.py:78]:
```suggestion
# COM: As the output below is generated by catalyst-cli, checking the correct order may cause flaky results.
```

**Pull request merged**